File: CookieClickerJetpack/app/src/main/java/com/example/cookieclickerjetpack/entity/Prefs.kt
package com.example.cookieclickerjetpack.entity



import android.content.Context
import com.example.cookieclickerjetpack.model.GameState
import com.example.cookieclickerjetpack.model.PerkStore

import org.json.JSONObject
import kotlin.math.roundToInt

class Prefs(ctx: Context) {
    private val sp = ctx.getSharedPreferences("cookie_prefs_compose", Context.MODE_PRIVATE)

    fun loadGame(): GameState {
        val s = sp.getString("game", null) ?: return GameState()
        val j = JSONObject(s)
        return GameState(
            score = j.optDouble("score", 0.0),
            cps = j.optDouble("cps", 0.0),
            baseTap = j.optInt("baseTap", 1),
            extraTap = j.optInt("extraTap", 0),
            prestigeLevel = j.optInt("prestigeLevel", 0)
        )
    }
    fun saveGame(gs: GameState) {
        val j = JSONObject()
        j.put("score", gs.score)
        j.put("cps", gs.cps)
        j.put("baseTap", gs.baseTap)
        j.put("extraTap", gs.extraTap)
        j.put("prestigeLevel", gs.prestigeLevel)
        sp.edit().putString("game", j.toString()).apply()
    }

    fun loadPerks(): PerkStore {
        val s = sp.getString("perks", null) ?: return PerkStore()
        val j = JSONObject(s)
        return PerkStore(
            points = j.optInt("points", 0),
            gprod = j.optInt("gprod", 0),
            crit = j.optInt("crit", 0),
            discount = j.optInt("discount", 0),
            tapTop = j.optInt("tapTop", 0)
        )
    }
    fun savePerks(ps: PerkStore) {
        val j = JSONObject()
        j.put("points", ps.points)
        j.put("gprod", ps.gprod)
        j.put("crit", ps.crit)
        j.put("discount", ps.discount)
        j.put("tapTop", ps.tapTop)
        sp.edit().putString("perks", j.toString()).apply()
    }
}

File: react_native_cookie_clicker/src/data/prefs.ts
// src/prefs.ts
import AsyncStorage from "@react-native-async-storage/async-storage";
import {PerkStore} from "../model/perk_store";
import {GameState} from "../model/game_state";

const KEY_GAME = "cookie_prefs_compose_game";
const KEY_PERKS = "cookie_prefs_compose_perks";

export const defaultGameState: GameState = {
    score: 0,
    cps: 0,
    baseTap: 1,
    extraTap: 0,
    prestigeLevel: 0,
};

export const defaultPerks: PerkStore = {
    points: 0,
    gprod: 0,
    crit: 0,
    discount: 0,
    tapTop: 0,
};

export async function loadGame(): Promise<GameState> {
    const s = await AsyncStorage.getItem(KEY_GAME);
    if (!s) return defaultGameState;
    try {
        const j = JSON.parse(s);
        return {
            score: Number(j.score ?? 0),
            cps: Number(j.cps ?? 0),
            baseTap: Number(j.baseTap ?? 1),
            extraTap: Number(j.extraTap ?? 0),
            prestigeLevel: Number(j.prestigeLevel ?? 0),
        };
    } catch {
        return defaultGameState;
    }
}

export async function saveGame(gs: GameState) {
    await AsyncStorage.setItem(KEY_GAME, JSON.stringify(gs));
}

export async function loadPerks(): Promise<PerkStore> {
    const s = await AsyncStorage.getItem(KEY_PERKS);
    if (!s) return defaultPerks;
    try {
        const j = JSON.parse(s);
        return {
            points: Number(j.points ?? 0),
            gprod: Number(j.gprod ?? 0),
            crit: Number(j.crit ?? 0),
            discount: Number(j.discount ?? 0),
            tapTop: Number(j.tapTop ?? 0),
        };
    } catch {
        return defaultPerks;
    }
}

export async function savePerks(ps: PerkStore) {
    await AsyncStorage.setItem(KEY_PERKS, JSON.stringify(ps));
}
